### PR TITLE
Globalmarks

### LINKED ISSUE #1692: Jump to Mark in Any file
* Click *thumbs-up* 👍 on this issue if you want it!
* Click *confused* 😕 on this issue if not having it makes VSCodeVim unusable. 

The VSCodeVim team prioritizes issues based on reaction count. 

--------

**BUG REPORT **:

**Environment**:

- **VSCode Version**: 1.12.1
- **VsCodeVim Version**: 0.7.1
- **OS**: 

**What happened**:
Marks do not work across files  using command 
`{A-Z}	go to mark {A-Z} in any file

**What did you expect to happen**: 
Jump to mark in another file

**How to reproduce it**:
Have a different file active, jumping marks do not do anything

### PR DESCRIPTION
**Pull request by jessewritescode**
**What this PR does / why we need it**:
Supports file marks.  I.e., just like normal marks, except when the mark is uppercase it should jump across files in addition to the cursor position.

**Which issue(s) this PR fixes**
Fixes #1692
Burrows heavily from #564 

**Special notes for your reviewer**:
This is my first pull request on this repo, so I'm not sure if I've missed something.  I was going to add tests, but I couldn't see any existing tests for marks, so I decided not to.

### CODE REVIEW

**Comment by jpoon** [comment]
Mind adding tests?

**Comment by jessewritescode** [comment]
Sorry for the delay.  Added some tests.

**Comment by jessewritescode** [comment]
@jpoon Do you need me to remove the status label or something to indicate the this pull request is ready for further review?

**Comment by jpoon** [comment]
There's something wrong with vscode + travis builds right now... https://github.com/microsoft/vscode/issues/75873#issuecomment-504296634

**Comment by SPascareli** [comment]
Why wasn't this merged? The issue still persists where marks doesn't work between files.

**Comment by jessewritescode** [comment]
> Why wasn't this merged? The issue still persists where marks doesn't work between files.

I'm not sure, I left it here hoping it may be of use.  If someone wants to review it, I will try to respond to any criticisms.

**Comment by focus7eleven** [comment]
waiting for this to be merged

**Comment by suominentoni** [comment]
Please merge this!

File: src/history/historyTracker.ts [hunk 1]
```diff
@@ -342,36 +344,87 @@ export class HistoryTracker {
...
+   * editor.
+   */
+  getLocalMarks(): IMark[] {
+    return [...this.currentHistoryStep.marks];
+  }
+
+  /**
+   * Gets all file marks.  I.e., marks that are shared among all editors.
+   */
+  getFileMarks(): IMark[] {
```
Comment (J-Fields) [inline src/history/historyTracker.ts]: I think `getGlobalMarks()` would be a more descriptive title

File: src/history/historyTracker.ts [hunk 2]
```diff
@@ -91,6 +92,11 @@ class HistoryStep {
    */
   marks: IMark[] = [];
 
+  /**
+   * "global" file marks. (when IMark.name is uppercase)
+   */
+  static fileMarks: IMark[] = [];
```
Comment (J-Fields) [inline src/history/historyTracker.ts]: As below, can we change this to `globalMarks`?

**Comment by jessewritescode** [comment]
> I'll need to take another closer look, but on first glance, this looks good. Thanks for the PR and your patience!

No worries, and thanks.  I should be able to address your comments in the next few days.

**Comment by J-Fields** [comment]
Sorry about the notification spam. This isn't working for me - it puts me into the right file, but not the right line/column.

Also, we ideally would like to associate capital marks with a document rather than a `TextEditor`, so that if the editor is closed, we can still jump to it. If we want to do that in a second pass, though, that would be fine.

**Comment by jessewritescode** [comment]
> Sorry about the notification spam. This isn't working for me - it puts me into the right file, but not the right line/column.
> 
> Also, we ideally would like to associate capital marks with a document rather than a `TextEditor`, so that if the editor is closed, we can still jump to it. If we want to do that in a second pass, though, that would be fine.

No worries, it might be a week or two until I can get back to this--But I see if I can figure it out.

I agree with your note about linking to a document.  I could look at doing that in a subsequent PR (first I need to familiarize myself with vscode's APIs)

**Comment by J-Fields** [comment]
Sounds good. If I have some extra time in the meantime (wishful thinking, maybe) I'll try to finish this off myself. If not, feel free to reach out!

**Comment by J-Fields** [comment]
@jessewritescode I think the problem was upstream, and has been fixed. I'll see if I can get this merged tonight.

**Comment by J-Fields** [comment]
Thanks @jessewritescode!

**Comment by jessewritescode** [comment]
@J-Fields Thanks!  Sorry, I was unresponsive in the last month--all the changes came in right as I was changing jobs.

In a month or so, I wouldn't mind picking up some more tasks from the backlog. :)

**Comment by J-Fields** [comment]
No trouble at all, hope you're enjoying the new gig!

>In a month or so, I wouldn't mind picking up some more tasks from the backlog. :)

That'd be awesome 🙂

**Pull request merged**